File: ruby/largest-series-product/largest_series_product.rb
class Series
  def initialize(input)
    raise ArgumentError if input =~ /\D/

    @series = input.chars.map(&:to_i)
  end

  def largest_product(size)
    return 1 if size.zero?
    raise ArgumentError if @series.empty? || size > @series.length

    @series.each_cons(size)
           .map { |digits| digits.reduce(&:*) }
           .max
  end
end

File: python/largest-series-product/largest_series_product.py
from functools import reduce


def product(nums):
    return reduce(lambda p, n: p * n, nums, 1)


def largest_product(series, size):
    if size > len(series):
        raise ValueError("span must not exceed string length")

    if series != "" and not series.isdecimal():
        raise ValueError('digits input must only contain digits')

    if size < 0:
        raise ValueError('span must not be negative')

    # max() throws ValueError on empty sequence
    # `series == ""` and `size == 0` edge cases are handled
    return max([
        product(map(int, series[i:i+size]))
        for i in range(len(series) - size + 1)
    ])
